### ISSUE TITLE
How to get vfile.data from the serializer?

### ISSUE THREAD

**Issue by nightire** [hashicorp/next-mdx-remote#270]
I'm using a remark plugin to extract all headings and store them in the `vfile.data`, but the problem is the serializer only returns the `vfile.data.matter` so it's impossible to get more from `vfile.data`:

<img width="656" alt="image" src="https://user-images.githubusercontent.com/1104018/165958658-46d4c88b-debc-4954-accb-dfc126773294.png">

So, what should I do to get the extra things from `vfile.data`?

**Comment by BRKalow** [comment]
We could pass along the vfile data in `scope`, what do you think about that?

**Comment by nightire** [comment]
> We could pass along the vfile data in `scope`, what do you think about that?

Yes, that would be great, thanks.

**Comment by kxxt** [comment]
> We could pass along the vfile data in `scope`, what do you think about that?

Maybe it's better not to expose the vfile data in `scope`. Personally I just want to extract some useful info from it, but I don't want to pass them to MDX. We can pass that data through a new variable.

**Comment by jesstelford** [comment]
Since `serialize` accepts a `VFileCompatible`, you're able to pass in your own vfile, then later extract the data yourself:

```
const file = new VFile('**some** awesome `content`');
const mdx = serialize(file);
// file.data has your remark plugin's data
```

**Comment by talatkuyuk** [comment]
> Since `serialize` accepts a `VFileCompatible`, you're able to pass in your own vfile, then later extract the data yourself:
> 
> ```js
> const file = new VFile('**some** awesome `content`');
> const mdx = serialize(file);
> // file.data has your remark plugin's data
> ```

You are right @jesstelford, but it is viable for only `pages` router since you can sent the data to client and inject it into MDX within the scope.  It is not possible to pass `file.data` into the MDX within the `scope` especially for the `app` router. So, the requested feature still can be useful.

**Comment by Moriarty47** [comment]
Any updates ?

**Comment by talatkuyuk** [comment]
I would like to inform that the [`next-mdx-remote-client`](https://github.com/hashicorp/next-mdx-remote/discussions/438#discussioncomment-8648430) supports copying data from `vfile.data` into `scope`.

```typescript
// ...

const options: SerializeOptions = {
  // ...
  vfileDataIntoScope: ["toc"], // the plugin "remark-flexible-toc" produces vfile.data.toc
  mdxOptions: {
    // ...
    remarkPlugins: [remarkFexibleToc],
  },
};

const mdxSource = await serialize({ source, options });
```

Now, the `scope` has `toc` namely table of contents, automatically, so you can use it in MDX document like:
```markdown
<TableOfContentComponent toc={toc} />
```

[Here](https://github.com/ipikuka/next-mdx-remote-client?tab=readme-ov-file#vfiledataintoscope) is the explanation. You can have a try it.
